File: C/Queue.c
//circular queue implementation in C using array

#include <stdio.h>
#define MAX 100

typedef struct Queue {
    int arr[MAX];
    int front, rear;
} Queue;

void initQueue(Queue* q) { q->front = q->rear = -1; }

int isEmpty(Queue* q) { return q->front == -1; }
int isFull(Queue* q) { return (q->rear + 1) % MAX == q->front; }

void enqueue(Queue* q, int val) {
    if (isFull(q)) {
        printf("Queue Overflow\n");
        return;
    }
    if (isEmpty(q)) q->front = q->rear = 0;
    else q->rear = (q->rear + 1) % MAX;
    q->arr[q->rear] = val;
}

int dequeue(Queue* q) {
    if (isEmpty(q)) {
        printf("Queue Underflow\n");
        return -1;
    }
    int val = q->arr[q->front];
    if (q->front == q->rear) q->front = q->rear = -1;
    else q->front = (q->front + 1) % MAX;
    return val;
}

int main() {
    Queue q;
    initQueue(&q);

    enqueue(&q, 10);
    enqueue(&q, 20);
    enqueue(&q, 30);

    printf("Dequeued: %d\n", dequeue(&q));
    printf("Dequeued: %d\n", dequeue(&q));

    return 0;
}

File: go/queue.go
package main

import "fmt"

type Queue struct {
	items []int
	front int
	rear  int
}

func NewQueue() *Queue {
	return &Queue{
		items: make([]int, 0),
		front: -1,
		rear:  -1,
	}
}

func (q *Queue) IsEmpty() bool {
	return q.front == -1
}

func (q *Queue) Enqueue(item int) {
	if q.IsEmpty() {
		q.front = 0
	}
	q.rear++
	q.items = append(q.items, item)
}

func (q *Queue) Dequeue() int {
	if q.IsEmpty() {
		return -1
	}
	
	item := q.items[q.front]
	q.front++
	
	if q.front > q.rear {
		q.front = -1
		q.rear = -1
	}
	
	return item
}

func (q *Queue) Display() {
	if q.IsEmpty() {
		fmt.Println("Queue is empty")
		return
	}
	
	for i := q.front; i <= q.rear; i++ {
		fmt.Print(q.items[i], " ")
	}
	fmt.Println()
}

func main() {
	queue := NewQueue()
	queue.Enqueue(10)
	queue.Enqueue(20)
	queue.Enqueue(30)
	fmt.Print("Queue: ")
	queue.Display()
	fmt.Println("Dequeued:", queue.Dequeue())
	fmt.Print("Queue after dequeue: ")
	queue.Display()
}
